paragraph/graph/graph.h

```diff
@@ -79,7 +79,12 @@ class Graph {
 
   // Checks if all instructions in the graph are reachable in PostOrder view
   bool IsConnected();
-  //
+
+  // Drops empty branches of the graph, meaning subroutines belonging to
+  // general purpose instructions in which all embedded instructions has zero
+  // seconds assigned
```

Comment:
There is a blacklist of what can't be dropped, i.e. non-call control flow, communication instructions, it's described in private method `Instruction::DropEmptyInnerSubroutines()`

Reply:
The implicit rule is for what cannot be converted is:
1) not part of communication instruction as they may be interpreted by simulator
2) not non-trivial control flow (not call), as we can drop some subroutines in While or Conditional and that makes graph invalid
3) everything in the subroutines to be deleted, including all recursively nested instructions, should lead to 0 seconds modeling. There might be non-trivial dependencies cut, but if all instructions are 0 cycle instructions that should not change the outcome of simulation)

I can change it to 0 seconds, 0 flops, and 0 bytes, but essentially the goal here is to cut instructions that only takes space in graph and cannot possibly change the outcome of simulation

Reply:
"0 seconds, 0 flops, and 0 bytes", that is probably the right thing to do. 

Suppose you have N delay instructions with non-zero seconds. After that there is a zero second delay. After that there is N more non-zero delay instructions. It is obvious that the purpose of the zero second delay was to reduce the number of dependencies from N^2 to 2N. So what will you do?

Reply:
I don't delete "empty" instructions on their own, I delete whole subroutines where all the instructions are empty, and only for trivial non-collective instructions. So if I have mix of zero-delay and non-zero delay instructions in subroutine, I won't delete any, as I don't analyze dependencies between instructions to decide what's safe to delete, and what not because of some non-trivial dependencies. It's a very basic coarse grain approach.